lib/auth/github.go

```diff
@@ -210,61 +203,157 @@ func (s *AuthServer) validateGithubAuthCallback(q url.Values) (*GithubAuthRespon
-	if len(logins) == 0 {
-		return trace.BadParameter(
+func (s *AuthServer) createWebSession(user services.User, sessionTTL time.Duration) (services.WebSession, error) {
+	session, err := s.NewWebSession(user.GetName())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Session expiry time is the same as the user expiry time.
+	session.SetExpiryTime(s.clock.Now().UTC().Add(sessionTTL))
+
+	// Bearer tokens expire quicker than the overall session time and need to be refreshed.
+	bearerTTL := utils.MinTTL(BearerTokenTTL, sessionTTL)
+	session.SetBearerTokenExpiryTime(s.clock.Now().UTC().Add(bearerTTL))
+
+	err = s.UpsertWebSession(user.GetName(), session)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return session, nil
+}
+
+func (s *AuthServer) createSessionCert(user services.User, sessionTTL time.Duration, publicKey []byte, compatibility string) ([]byte, []byte, error) {
+	roles, err := services.FetchRoles(user.GetRoles(), s.Access, user.GetTraits())
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	certs, err := s.generateUserCert(certRequest{
+		user:          user,
+		roles:         roles,
+		ttl:           sessionTTL,
+		publicKey:     publicKey,
+		compatibility: compatibility,
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return certs.ssh, certs.tls, nil
+}
+
+// createUserParams is a set of parameters used to create a user for an\
+// external identity provider.
+type createUserParams struct {
+	// connectorName is the name of the connector for the identity provider.
+	connectorName string
+
+	// username is the Teleport user name .
+	username string
+
+	// logins is the list of *nix logins.
+	logins []string
+
+	// kubeGroups is the list of Kubernetes this user belongs to.
+	kubeGroups []string
+
+	// roles is the list of roles this user is assigned to.
+	roles []string
+
+	// traits is the list of traits for this user.
+	traits map[string][]string
+
+	// sessionTTL is how long this session will last.
+	sessionTTL time.Duration
+}
+
+func (s *AuthServer) calculateGithubUser(connector services.GithubConnector, claims *services.GithubClaims, request *services.GithubAuthRequest) (*createUserParams, error) {
+	var p createUserParams
+
+	p.connectorName = connector.GetName()
+	p.username = claims.Username
+
+	// Calculate logins, kubegroups, roles, and traits.
+	p.logins, p.kubeGroups = connector.MapClaims(*claims)
+	if len(p.logins) == 0 {
+		return nil, trace.BadParameter(
 			"user %q does not belong to any teams configured in %q connector",
 			claims.Username, connector.GetName())
 	}
+	p.roles = modules.GetModules().RolesFromLogins(p.logins)
+	p.traits = modules.GetModules().TraitsFromLogins(p.logins, p.kubeGroups)
+
+	// Pick smaller for role: session TTL from role or requested TTL.
+	roles, err := services.FetchRoles(p.roles, s.Access, p.traits)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	roleTTL := roles.AdjustSessionTTL(defaults.CertDuration)
```

Comment:
should be Max duration, not default cert duration, so we can process larger than default values